Prep changes for mute user and some minor bug fixes

abhijeetbodas2001 opened the pull request:
Refactoring part of #16915

timabbott commented:
This is really nice work!  I merged the first few commits as the series ending with e8b4902c8ae747251babefd61ab755dd4d1d84f0.  I think for `.muting_enabled`, we should go with `.excludes_muted_topics`?  I think `consider_muted_topics` is too ambiguous, as it suggests we might show messages in muted topics sometimes.  (Or is the intent to indicate that, since we do show mentioned messages anyway?)

abhijeetbodas2001 commented:
I had thought of `excludes_muted_topic_messages`, but I feel that's a bit too long. If we can treat the "messages" part as implicit here, I think `excludes_muted_topics` is definitely a better name.

abhijeetbodas2001 commented:
Changed to use `excludes_muted_topics`. This is ready for another review. Apart from the re-renaming, I also made some tiny changes-
<details><summary>diff</summary>
<p>

```bash
(zulip-py3-venv) abhijeetbodas2001@DESKTOP-TLMTHQ2:~/zulip$ git diff
diff --git a/frontend_tests/node_tests/message_list_data.js b/frontend_tests/node_tests/message_list_data.js
index 25cfca0058..38879e6bce 100644
--- a/frontend_tests/node_tests/message_list_data.js
+++ b/frontend_tests/node_tests/message_list_data.js
@@ -125,7 +125,7 @@ run_test("muting", () => {
     ];

     // `exclude_topic_muted_messages` should skip filtering
-    // messages if `consider_mutes` is false.
+    // messages if `excludes_muted_topics` is false.
     let is_topic_muted_calls = 0;

     with_field(
@@ -135,8 +135,9 @@ run_test("muting", () => {
             is_topic_muted_calls = is_topic_muted_calls + 1;
         },
         () => {
-            mld.exclude_topic_muted_messages(msgs);
+            const res = mld.exclude_topic_muted_messages(msgs);
             assert.equal(is_topic_muted_calls, 0);
+            assert.deepEqual(res, msgs);
         },
     );

@@ -150,7 +151,7 @@ run_test("muting", () => {
     ]);

     // MessageListData methods should always attempt to filter messages,
-    // and update `_all_items` when `consider_mutes` is true.
+    // and update `_all_items` when `excludes_muted_topics` is true.
     mld = new MessageListData({
         excludes_muted_topics: true,
         filter: undefined,
```

</p>
</details>

showell commented:
For the new function `exclude_topic_muted_messages` it sounds like the function mutates a list or does something with more side effects, when it's actually returning either the original list unmodified (when topic-muting does not apply) or a new list.

So I think I would prefer something more like `messages_filtered_for_topic_mutes` or something.

If we can't agree on a significantly better name here, I think it's fine to just add a little comment to the function that says it returns a list and doesn't mutate anything.

showell commented on static/js/message_list_data.js:
```diff
@@ -170,15 +170,25 @@ class MessageListData {
 
+    exclude_topic_muted_messages(messages) {
+        if (!this.excludes_muted_topics) {
+            return messages;
```
If we wanted to be extra defensive here, we could return `[...messages]`, which would mean the function consistently returns a new list.  I doubt we would ever mutate the list that gets passed in to this function in a harmful way, but it's also very cheap to copy the list here.

showell replied:
(I also propose to rename this function in the general comments.)

timabbott commented:
I agree with Steve: `messages_filtered_for_topic_mutes` would be a great name for that function.

abhijeetbodas2001 commented:
In the last push, I renamed the function to `messages_filtered_for_topic_mutes`, and changed it so that it always returns a new list, as mentioned above in [Steve's comment](https://github.com/zulip/zulip/pull/17215#discussion_r571479696).
